# TranSurVeyor

## Compiling

To be compiled, TranSurveyor requires g++ 4.7.2.

htslib 1.7 is included in the source, zipped. First of all, you should build it by using the provided script
```
./build_htslib.sh
```
If htslib does not build correctly, please refer to https://github.com/samtools/htslib

Then, run

```
cmake . && make
```

## Preparing the reference

The reference fasta file should be indexed by both bwa and samtools. For example, assuming the file is hg19.fa, you should run
```
bwa index hg19.fa
samtools faidx hg19.fa
```

## Preparing the BAM file

TranSurveyor has currently only been tested using BAM files generated by BWA MEM, therefore we recommend its usage. 
It should also be run through Picard FixMateInformation (http://broadinstitute.github.io/picard/command-line-overview.html#FixMateInformation);
in particular, it should have the MQ and MC tags.
Finally, the file should be sorted and indexed ad usual using samtools.

Supposing file.bam is the file resulting from the alignment:
```
java -jar picard.jar FixMateInformation I=file.bam
samtools sort file.bam > sorted.bam
samtools index sorted.bam
```

## Running 

Once the c++ code is compiled, TranSurveyor can be run. Python and libraries NumPy (http://www.numpy.org/), PyFaidx (https://github.com/mdshw5/pyfaidx) and PySam (https://github.com/pysam-developers/pysam) are required. Python 2.7, NumPy 1.10, PyFaidx 0.4 and PySam 0.12 are the recommended versions.

The bare minimum command for running TranSurveyor is 
```
python surveyor.py /path/to/bamfile /an/empty/working/directory /path/to/reference/fasta
```

Other parameters which may be important are the number of threads and the location of the bwa and samtools executable.
```
python surveyor.py /path/to/bamfile /an/empty/working/directory	/path/to/reference/fasta --threads 40 --samtools /path/to/samtools --bwa /path/to/bwa
```

After TranSurveyor has been successfully run, the calls can be retrieved with the command
```
./filter /path/to/working/directory
```

## Notes

TranSurveyor is a very young software, and it has only been used by its authours on a handful of datasets. 
If you should encounter problems, please write to e0011356@u.nus.edu ; we will do our best to both improve/fix both the software and the documentation.

